src/invidious/playlists.cr

```diff
@@ -437,17 +437,45 @@ def get_playlist_videos(db, playlist, offset, locale = nil, continuation = nil)
     db.query_all("SELECT * FROM playlist_videos WHERE plid = $1 ORDER BY array_position($2, index) LIMIT 100 OFFSET $3",
       playlist.id, playlist.index, offset, as: PlaylistVideo)
   else
-    if offset >= 100
-      # Normalize offset to match youtube's behavior (100 videos chunck per request)
-      offset = (offset / 100).to_i64 * 100_i64
+    videos = [] of PlaylistVideo
+
+    original_offset = offset
+    until videos.size >= 100 || videos.size == playlist.video_count || offset >= playlist.video_count
+      if offset >= 100
+        # Normalize offset to match youtube's behavior (100 videos chunck per request)
+        normalized_offset = (offset / 100).to_i64 * 100_i64
+        ctoken = produce_playlist_continuation(playlist.id, normalized_offset)
+        initial_data = request_youtube_api_browse(ctoken)
+      else
+        initial_data = request_youtube_api_browse("VL" + playlist.id, params: "")
+      end
```

Comment:
I'm wondering why you put that chunk of code in a loop. This retrieves a chunk of 100 videos, so calling it 100 times is extremely expensive computation and bandwidth-wise.

Reply:
Somehow I missed that during my review... 

Effectively that loop only goes through one iteration since at least one of it's conditions will become true no matter what:

If all 100 videos for the current paged are fetched and fulfilled? Loop ends

If all videos for the playlist are fetched? Loop ends.

Etc 

However, I think I do recall a few cases where some Invidious playlist pages doesn't show the full 100 videos while still having more pages to access.  

I'm assuming this is what @diogorac meant by:
> Also made sure that the playlist is always filled with videos if possible

So the loop is likely just an attempt to resolve that issue.

Reply:
Unless I'm missing something, it's only being called until we find the corresponding videoID (if we're missing an offset), have "enough" videos on the list and/or we reach the end of the play list. 
``` until videos.size >= 100 || videos.size == playlist.video_count || offset >= playlist.video_count```

I hadn't looked into the ```request_youtube_api_browse```, and just worked with what it gave me. But now that I did, I see no use for this normalized offset. It's just making the results to be limited to pages from 0-99/100-199/etc, where the requested offset video is not the first one. Which means, we have to shift towards it, and then request more videos to fill the playlist in front of us. Otherwise, if it's video number 99, we won't have any following videos.
Without the normalized offset, it will return a playlist exactly from the offset + 99 videos, which is more suited. And it would make the loop meaningless... However...

The main reason I added it, was because of URLs where there is no offset. eg:
https://yewtu.be/playlist?list=PLNYlAz8KKiVl6HigSczBEHDdZhb5GdrOB
Any of those links have no ```index```. Now one solution might be to add an index to the html. But it has happened to me before, when I click some similar youtube link, that gets redirected to my invidious instance and I face the same issue.
If there is no offset, we have to look for our video in each 100-page-chunk, and when we find it, we fill our playlist with more videos (again same 99th video example.)

Reply:
In that case wouldn't it be better to just request InnerTube's `youtubei/v1/next` endpoint? For instance, if you request with these post data — and everything else that InnerTube requires for a response: 
`videoId": "xpCj7oaWPnY", "playlistId": "PLsNB4peY6C6KhUZiP_YogoCH4gsMccbXN` 

The following JSON would then be returned:
```json
...
      "playlist": {
        "playlist": {
          ...
          "currentIndex": 13,
          "playlistId": "PLsNB4peY6C6KhUZiP_YogoCH4gsMccbXN",
          "totalVideos": 2371
        }
     }
```
Which contains the index of the current video within the requested playlist.

Reply:
Oh cool I didn't know about this API. That simplifies things.

 I added it and increased the lookback to 100 and max videos to 200.

Reply:
> I hadn't looked into the `request_youtube_api_browse`, and just worked with what it gave me. But now that I did, I see no use for this normalized offset. It's just making the results to be limited to pages from 0-99/100-199/etc, where the requested offset video is not the first one.

This bit of code was made for playlist pages. You can still create a ctoken with the video ID you want, youtube will always return this video and the next 99 ones.

> Which means, we have to shift towards it, and then request more videos to fill the playlist in front of us. Otherwise, if it's video number 99, we won't have any following videos.

Not necessarily. If you're playing the 205th video in a playlist, you can always craft a ctoken with offset = 155, so the API will give you all the videos from 155 to 254, inclusive.

> The main reason I added it, was because of URLs where there is no offset. eg:
> https://yewtu.be/playlist?list=PLNYlAz8KKiVl6HigSczBEHDdZhb5GdrOB
> Any of those links have no `index`. Now one solution might be to add an index to the html. (which I did) But it has happened to me before, when I click some similar youtube link, that gets redirected to my invidious instance and I face the same issue.

we shouldn't need an `index=` parameter. We know what the current video is, we already have a list of videos, so we can do everything on the back-end side.

Reply:
>This bit of code was made for playlist pages. You can still create a ctoken with the video ID you want, youtube will always return this video and the next 99 ones.
>Not necessarily. If you're playing the 205th video in a playlist, you can always craft a ctoken with offset = 155, so the API will give you all the videos from 155 to 254, inclusive.

Yeah I understood that later, and that's why I changed it so it crafts the ctoken with the specific offset. 

>we shouldn't need an index= parameter. We know what the current video is, we already have a list of videos, so we can do everything on the back-end side.

And we don't need. But by having it, we makes things smoother on the backend. That way we don't have to search for the index with the```v1/next``` request.

But maybe I'm missing something?

Reply:
> And we don't need. But by having it, we make things smoother on the backend. That way we don't have to search for the index with the`v1/next` request.
> 
> But maybe I'm missing something?

Nope, you're right :) That index thing is a good idea, and it looks like youtube does it, too...